In [6]:
# Install otter-grader to import
pip install otter-grader

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 162.1/162.1 KB 5.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.5/77.5 KB 11.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 307.2/307.2 KB 21.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 KB 15.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 767.1/767.1 KB 25.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 103.2/103.2 KB 15.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 44.2 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 572.7/572.7 KB 58.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.1/52.1 KB 8.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.5/84.5 KB 16.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.3/47.3 KB 11.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━

In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("hwk1-task2-corals.ipynb")

# Task 2: Exploring coral diversity data

## Instructions 

- First, update the following cell to have a link to *your* Homework 1 GitHub repository:


**UPDATE THIS LINK:**

[https://github.com/nmoraescommit/eds220-hwk1.git]()

- This task covers topics on the lesson on [subsetting `pandas.DataFrames`](https://meds-eds-220.github.io/MEDS-eds-220-course/book/chapters/lesson-3-pandas-subsetting/lesson-3-pandas-subsetting.html). The best strategy to solve it is to get together with your classmates and work on it together after that lecture.

- Review the [complete rubric for this task](https://docs.google.com/document/d/131OnUDOLpOl94SlaMU9nGqT998ABrjFS/edit?usp=sharing&ouid=111507336322611936333&rtpof=true&sd=true) before starting.

- **Make at least 5 commits at moments you think are adequate.** We'll check your repository and view the commit history.

- **Add comments for all your code.** Err on the side of commenting too much for now. Comments should follow best practices.

- **Do not update the top cell with the `otter` import**, this is used internally for grading.



<!-- BEGIN QUESTION -->


## 1
For this task we are going to use data about  [Western Indian Ocean Coral Diversity (McClanahan, 2023)](https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1K35S3H) from the Knowledge Network for Biocomplexity (KNB) data repository. 

a. Navigate to the dataset's link and explore the data package. 
 
b. Use the next markdown cell to add a brief description of the dataset, including a citation, date of access, and a link to the archive. 


1. Description:
2. Citation: Tim McClanahan. (2023). Western Indian Ocean Coral Diversity. Knowledge Network for Biocomplexity. doi:10.5063/F1K35S3H.
3. Date of Access: 10/07/2024
4. Archive Link: https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1K35S3H#urn%3Auuid%3Ae7d12a5a-f1a9-4a3e-aaa5-5222f67e799d

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 2
In the next code cell, import the `pandas` package using standard abbreviation in a code cell.

Read in the data from the URL using the `pd.read_csv` function and store it as `coral_div` like this:

```
# Read in data
coral_div = pd.read_csv('the URL goes here')
```

In [3]:
# import packages
import pandas as pd

coral_div = pd.read_csv('https://knb.ecoinformatics.org/knb/d1/mn/v2/object/urn%3Auuid%3Ae7d12a5a-f1a9-4a3e-aaa5-5222f67e799d')

<!-- END QUESTION -->

## 3

Use attributes and methods of the `coral_div` data frame to obtain the following information about it:

Use the code cells marked a-e. For each, add a comment explaining what is the ouput.

a. how many rows and columns does it have? Store your answers in the `rows` and `columns` variables. 

In [21]:
# a

# use following code to print to answer

print("The data frame has", coral_div.shape[0], "rows and", coral_div.shape[1], "columns.", sep = " ")

# store the number of rows and columns of coral_div in appropriate variables

rows = coral_div.shape[0]
columns = coral_div.shape[1]

The data frame has 575 rows and 42 columns.


b. What are the column names? Store your answer in the `coral_div_columns` variable.

In [23]:
# b

#  use following code to print to answer

print("The column labels are as follows:", coral_div.columns)

# store the column names in coral_div_columns variable

coral_div_columns = coral_div.columns

The column labels are as follows: Index(['PU_ID', 'PU_Lat', 'PU_Lon', 'Number_of_genera', 'Country', 'Ecoregion',
       'Management', 'Habitat', 'Observer', 'Depth', 'Coralcover',
       'SSTRateofRise', 'PARmax', 'Calcite', 'Dis_oxygen', 'mean.waveNRJ',
       'Salinity_mean', 'andrello_reef_value', 'andrello_nutrients',
       'mean.npp', 'TT_market_hrs', 'TT_pop_hrs', 'cumDHW', 'SST_Kurtosis',
       'SST_Median', 'ChlorA_median', 'andrello_sediments', 'Current_vel_mean',
       'Diff_attenuation', 'Grav_NP', 'Grav_NC', 'PH', 'SST_bimodality',
       'SST_Skewness', 'ClimateStressModel', 'Netflow', 'Indegree',
       'Outdegree', 'Retention', 'min_yr', 'max_yr', 'no_replicates'],
      dtype='object')


c. what are the data types of the columns? Store your answer in the `coral_div_types`.

In [32]:
# c

#  use following code to print to answer

print("The column data types in the coral_div data frame is as follows:\n",coral_div.dtypes)

# store the column names in coral_div_types variable

coral_div_types = coral_div.dtypes

The column data types in the coral_div data frame is as follows:
 PU_ID                    int64
PU_Lat                 float64
PU_Lon                 float64
Number_of_genera         int64
Country                 object
Ecoregion               object
Management              object
Habitat                 object
Observer                object
Depth                  float64
Coralcover               int64
SSTRateofRise          float64
PARmax                 float64
Calcite                float64
Dis_oxygen             float64
mean.waveNRJ           float64
Salinity_mean          float64
andrello_reef_value      int64
andrello_nutrients     float64
mean.npp               float64
TT_market_hrs          float64
TT_pop_hrs             float64
cumDHW                 float64
SST_Kurtosis           float64
SST_Median             float64
ChlorA_median          float64
andrello_sediments     float64
Current_vel_mean       float64
Diff_attenuation       float64
Grav_NP                float64
Grav

d. display the first 8 rows with all the column names. For this one, you will need to udpate a `pandas` option (HINT: check the first discussion section). Store your answer in the `first_8_rows` variable.

In [43]:
# d

# use the method head() to store first 8 rows with all column names

first_8_rows = coral_div.head(8)

In [44]:
# View the dataframe 
first_8_rows

PU_ID     PU_Lat    PU_Lon  Number_of_genera   Country  \
0      7  39.076711 -5.490011                35  Tanzania   
1      7  39.076711 -5.490011                22  Tanzania   
2      7  39.076711 -5.490011                25  Tanzania   
3      7  39.076711 -5.490011                26  Tanzania   
4     10  39.054253 -5.512514                27  Tanzania   
5     10  39.054253 -5.512514                29  Tanzania   
6     15  39.054253 -5.535017                19  Tanzania   
7     39  39.009337 -5.647518                28  Tanzania   

                  Ecoregion   Management           Habitat Observer  Depth  \
0  East African Coral Coast  UnfishedLow             Crest      Tim   11.0   
1  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    2.0   
2  East African Coral Coast  UnfishedLow             Slope  Nyawira    4.0   
3  East African Coral Coast  UnfishedLow             Slope  Nyawira    3.0   
4  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    1.5   
5  East African Coral Coast  UnfishedLow  Lagoon/Back reef      Tim    1.5   
6  East African Coral Coast   Restricted             Crest      Tim    8.0   
7  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    3.0   

   ...  SST_bimodality  SST_Skewness  ClimateStressModel   Netflow  Indegree  \
0  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
1  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
2  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
3  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
4  ...        0.461297     -0.134548            0.629000  0.011721    112.75   
5  ...        0.461297     -0.134548            0.629000  0.011721    112.75   
6  ...        0.461297     -0.134548            0.595600  0.011721    112.75   
7  ...        0.464874     -0.158601            0.626000 -0.063087     96.50   

   Outdegree  Retention  min_yr  max_yr  no_replicates  
0      74.75      14.00    2019    2019              1  
1      74.75      14.00    2010    2015              2  
2      74.75      14.00    2015    2015              1  
3      74.75      14.00    2016    2019              2  
4      51.50     183.75    2010    2019              3  
5      51.50     183.75    2015    2015              1  
6      51.50     183.75    2019    2019              1  
7      40.75      35.75    2015    2019              3  

[8 rows x 42 columns]

<!-- BEGIN QUESTION -->

e. some other inital information you are interested in. 

In [46]:
# e 

# basic statistical information

coral_div_stats = coral_div.describe()

# view coral_div_stats

coral_div_stats

PU_ID      PU_Lat      PU_Lon  Number_of_genera       Depth  \
count    575.000000  575.000000  575.000000        575.000000  575.000000   
mean    3536.620870   43.976240  -12.633028         20.182609    6.253192   
std     2345.438008    6.549987    6.972048          6.568299    4.501193   
min        7.000000   32.878336  -27.011417          3.000000    0.500000   
25%     1413.500000   39.413579  -20.006995         16.000000    2.000000   
50%     4002.000000   40.671221  -12.827030         20.000000    5.000000   
75%     5182.000000   48.497793   -5.523765         24.000000    9.500000   
max    11683.000000   57.806585   -1.745137         40.000000   29.000000   

       Coralcover  SSTRateofRise      PARmax     Calcite  Dis_oxygen  ...  \
count  575.000000     575.000000  575.000000  575.000000  575.000000  ...   
mean    38.340870       0.017559   55.558422    0.003622    4.684780  ...   
std     10.586395       0.006552    2.118008    0.007026    0.108744  ...   
min     10.000000      -0.000463   48.774500    0.000058    4.373000  ...   
25%     31.000000       0.015466   54.117250    0.000272    4.626000  ...   
50%     39.000000       0.019233   55.523998    0.000906    4.711000  ...   
75%     45.000000       0.021978   57.133999    0.002866    4.755500  ...   
max     60.000000       0.027391   62.193001    0.055976    4.850250  ...   

       SST_bimodality  SST_Skewness  ClimateStressModel     Netflow  \
count      575.000000    575.000000          575.000000  575.000000   
mean         0.529946     -0.071522            0.549405    0.203586   
std          0.047659      0.127463            0.151452    0.344229   
min          0.416042     -0.297088            0.081800   -0.576851   
25%          0.488628     -0.168714            0.470100   -0.023227   
50%          0.547940     -0.095642            0.579325    0.233054   
75%          0.570981      0.018935            0.656300    0.509071   
max          0.598189      0.304314            0.813600    0.815635   

         Indegree   Outdegree    Retention       min_yr       max_yr  \
count  575.000000  575.000000   575.000000   575.000000   575.000000   
mean    51.882609   55.963043    91.107391  2013.165217  2013.923478   
std     37.238322   27.701567   147.657638     4.693628     4.638243   
min      2.500000   10.500000     0.000000  1998.000000  1998.000000   
25%     21.250000   35.625000    11.750000  2009.000000  2010.000000   
50%     43.500000   54.250000    47.750000  2014.000000  2015.000000   
75%     81.250000   77.000000   109.000000  2016.000000  2018.000000   
max    146.750000  136.000000  1561.500000  2022.000000  2022.000000   

       no_replicates  
count     575.000000  
mean        1.716522  
std         1.524934  
min         1.000000  
25%         1.000000  
50%         1.000000  
75%         2.000000  
max        19.000000  

[8 rows x 37 columns]

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 4

Some of the columns have data type `int64` and some have data type `float64`. What is the difference between these `int64` and `float64`?


While both data types can store 64 bits, the difference between int64 and float64 data is that int64 only stores whole numbers and float64 stores values that have decimal places.

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 5

The `Country` column indicates the country where the coral reef is located. Explain, in your own words, what is the ouput of `coral_div['Country']` and `coral_div['Country'].unique()`. 

In [47]:
print(coral_div['Country'])

print(coral_div['Country'].unique())

0          Tanzania
1          Tanzania
2          Tanzania
3          Tanzania
4          Tanzania
           ...     
570    South Africa
571    South Africa
572    South Africa
573    South Africa
574    South Africa
Name: Country, Length: 575, dtype: object
['Tanzania' 'Mozambique' 'Kenya' 'Reunion' 'Comoros' 'Mayotte'
 'French Southern Territories' 'Madagascar' 'Seychelles' 'South Africa'
 'Mauritius']


The output of `coral_div['Country']` is the column containing the names of the countries linked to each of the 573 observations, weheras the output of `coral_div['Country'].unique()` is the unique country names in the column.

<!-- END QUESTION -->

## 6

Check the dataset's [metadata file README_md.doc](https://knb.ecoinformatics.org/view/doi:10.5063/F1K35S3H) to find which column has information about who made an observation. How many observers collected this data? HINT: Check this [method for `pandas.DataFrame`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nunique.html). Store your answer in the `unique_observers` variable.

In [54]:
unique_observers = coral_div['Observer'].nunique()

# print statement detailing code

print("The number of observers are:", unique_observers)

The number of unique observers are: 16 .


## 7

The `Coralcover` column indicates the percent of hard coral cover in each 6 km reef cell surveyed. 

<!-- BEGIN QUESTION -->

a)  Describe, in detail,  what kind of object `coral_div['Coralcover']>= 50` is.

The object is a boolean. It returns a binary option, of true or false, in the context of the data in the column of Coralcover.

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

b) What data is `coral_div[coral_div['Coralcover'] >= 50]` selecting?

It is selecting the data in the Coralcover column, and returning true if it is greater than or equal to 50, otherwise it returns false.

<!-- END QUESTION -->

## 8

Use code to find if there are there any sites with a 30% coral cover in Tanzania. Store your answer in the `coral_cover` variable.

In [64]:
coral_cover = coral_div[coral_div['Coralcover'] == 30]

In [65]:
# View the dataframe
coral_cover

PU_ID     PU_Lat     PU_Lon  Number_of_genera     Country  \
63     693  39.391121  -5.782490                10    Tanzania   
64     697  39.391121  -5.804983                22    Tanzania   
122    867  39.368664  -4.656797                23       Kenya   
123    867  39.368664  -4.656797                26       Kenya   
124    868  39.391121  -4.656797                24       Kenya   
125    868  39.391121  -4.656797                20       Kenya   
126    868  39.391121  -4.656797                22       Kenya   
127    872  39.436037  -4.679331                22       Kenya   
139   1186  55.246386 -21.012018                19     Reunion   
210   2377  49.968784 -15.717807                15  Madagascar   
341   4294  39.211458  -4.814515                37    Tanzania   
342   4299  39.233916  -4.837043                26    Tanzania   
343   4302  39.211458  -4.859570                31    Tanzania   
344   4303  39.301290  -4.859570                23    Tanzania   
377   4906  39.660616  -4.972195                33    Tanzania   
388   5049  39.615700  -5.490011                14    Tanzania   
389   5050  39.638158  -5.490011                18    Tanzania   

                           Ecoregion   Management           Habitat  Observer  \
63          East African Coral Coast       Fished  Lagoon/Back reef   Nyawira   
64          East African Coral Coast       Fished  Lagoon/Back reef   Nyawira   
122         East African Coral Coast  UnfishedLow  Lagoon/Back reef   Nyawira   
123         East African Coral Coast  UnfishedLow  Lagoon/Back reef       Tim   
124         East African Coral Coast  UnfishedLow  Lagoon/Back reef   Nyawira   
125         East African Coral Coast  UnfishedLow  Lagoon/Back reef   Nyawira   
126         East African Coral Coast  UnfishedLow  Lagoon/Back reef       Tim   
127         East African Coral Coast   Restricted             Slope       Tim   
139                Mascarene Islands   Restricted               NaN  Mireille   
210  Western and Northern Madagascar       Fished               NaN       Tim   
341         East African Coral Coast       Fished             Slope   Nyawira   
342         East African Coral Coast       Fished             Slope   Nyawira   
343         East African Coral Coast       Fished             Slope   Nyawira   
344         East African Coral Coast       Fished             Slope   Nyawira   
377         East African Coral Coast       Fished  Lagoon/Back reef   Nyawira   
388         East African Coral Coast       Fished  Lagoon/Back reef   Nyawira   
389         East African Coral Coast       Fished  Lagoon/Back reef   Nyawira   

         Depth  ...  SST_bimodality  SST_Skewness  ClimateStressModel  \
63    5.000000  ...        0.490512     -0.101724            0.655600   
64    9.166667  ...        0.490769     -0.100365            0.683050   
122   2.666667  ...        0.448886     -0.036279            0.727025   
123   1.000000  ...        0.448886     -0.036279            0.727025   
124   4.000000  ...        0.450158     -0.035417            0.723350   
125   3.187500  ...        0.450158     -0.035417            0.723350   
126   4.000000  ...        0.450158     -0.035417            0.723350   
127  29.000000  ...        0.451827     -0.035988            0.628700   
139   2.500000  ...        0.568952     -0.009900            0.212200   
210   8.500000  ...        0.596164     -0.153239            0.470300   
341  11.000000  ...        0.451293     -0.048992            0.608900   
342  12.500000  ...        0.451293     -0.048992            0.601725   
343   7.000000  ...        0.452402     -0.050477            0.595200   
344  11.000000  ...        0.453806     -0.048007            0.602300   
377   9.000000  ...        0.488396     -0.074229            0.671900   
388   4.000000  ...        0.489921     -0.109523            0.657000   
389   6.000000  ...        0.489921     -0.109523            0.618825   

      Netflow  Indegree  Outdegree  Retent

## 9

We are interested in looking at the PH of sea water (`PH` column),  mean salinity (`Salinity_mean` column), and depth of survey location in meters (`Depth` column). Subset the data frame by selecting only these column. Store your answer in the `filtered_data` variable

In [67]:
filtered_data = coral_div[["PH", "Salinity_mean","Depth"]]

In [68]:
# View the dataframe
filtered_data

PH  Salinity_mean  Depth
0    8.2175      35.167646  11.00
1    8.2175      35.167646   2.00
2    8.2175      35.167646   4.00
3    8.2175      35.167646   3.00
4    8.2180      35.102688   1.50
..      ...            ...    ...
570  8.2400      35.301208  15.50
571  8.2400      35.301208  12.00
572  8.2400      35.301208  12.60
573  8.2400      35.301208  13.00
574  8.2400      35.301208  13.47

[575 rows x 3 columns]

## 10
The `max_yr` column indicates the year of last sampling at a given site. Select the PH, salinity, and depth for sites where sampling ended in 2017. Store your answer in the `sampling_2017` variable. 


In [72]:
sampling_2017 = filtered_data[coral_div['max_yr'] <= 2017]

In [73]:
# View the dataframe 
sampling_2017

PH  Salinity_mean      Depth
1    8.2175      35.167646   2.000000
2    8.2175      35.167646   4.000000
5    8.2180      35.102688   1.500000
8    8.2185      35.106329   6.500000
10   8.2170      35.105648   2.000000
..      ...            ...        ...
569  8.2400      35.301208  13.833333
570  8.2400      35.301208  15.500000
571  8.2400      35.301208  12.000000
572  8.2400      35.301208  12.600000
574  8.2400      35.301208  13.470000

[429 rows x 3 columns]

## 11
Use `iloc` to select rows 95 throuhg 100 (including 100) in the 17th, 20th, and 28th columns and assign these values to a new variable called `subset`. View the new `subset` dataframe.

HINT: the ouput columns should be `Salinity_mean`, `mean.npp` and `Current_vel_mean`. Make sure you are indexing from 0.

In [ ]:
subset = ...

In [ ]:
# View the dataframe
subset

<!-- BEGIN QUESTION -->

## 12 

Run `subset.isna()`. Explain what this code does and what the output represents.

_Type your answer here, replacing this text._

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

## 13
Run `subset.isna().any()`. Explain in your own words what this code does and what the output represents. HINT: check the [documentation of `pandas.Sries.any`](https://pandas.pydata.org/docs/reference/api/pandas.Series.any.html).


_Type your answer here, replacing this text._

<!-- END QUESTION -->

## 14 

Select the rows in `subset` that have NA in the `Current_vel_mean` column. Store your answer in the `subset_na` variable.

In [ ]:
subset_na = ...

In [ ]:
# View the dataframe 
subset_na